packages/paymasters/contracts/PermitERC20UniswapV3Paymaster.sol

```diff
@@ -91,25 +96,88 @@ contract PermitERC20UniswapV3Paymaster is BasePaymaster, ERC2771Recipient {
         gasUsedByPost = _gasUsedByPost;
     }
 
-    function _calculatePreCharge(
-        GsnTypes.RelayRequest calldata relayRequest,
-        uint256 maxPossibleGas)
-    internal
+    function setPaymasterFee(uint256 _paymasterFee) public onlyOwner {
+        paymasterFee = _paymasterFee;
+    }
+
+    function setMinHubBalance(uint256 _minHubBalance) public onlyOwner {
+        minHubBalance = _minHubBalance;
+    }
+
+    function setTargetHubBalance(uint256 _targetHubBalance) public onlyOwner {
+        targetHubBalance = _targetHubBalance;
+    }
+
+    function setMinWithdrawalAmount(uint256 _minWithdrawalAmount) public onlyOwner {
+        minWithdrawalAmount = _minWithdrawalAmount;
+    }
+
+    function setUniswap(ISwapRouter _uniswap) public onlyOwner {
+        uniswap = _uniswap;
+    }
+
+    function setTokens(
+        IERC20Metadata[] memory _tokens,
+        IChainlinkOracle[] memory _priceFeeds,
+        string[] memory _permitMethodSignatures,
+        uint24[] memory _poolFees) public onlyOwner {
+        tokens = _tokens;
+        // allow uniswap to transfer from paymaster balance
+        for (uint256 i = 0; i < tokens.length; i++) {
+            TokenSwapData memory data;
+            IERC20Metadata token = tokens[i];
+            token.approve(address(uniswap), type(uint256).max);
+            data.priceDivisor = 10 ** uint256(_priceFeeds[i].decimals() + token.decimals());
+            data.priceFeed = _priceFeeds[i];
+            data.permitMethodSignature = bytes4(keccak256(bytes(_permitMethodSignatures[i])));
```

Comment:
I would vote for accepting `bytes4`. I don't see any benefit in keccak-ing on-chain.

Reply:
This is your own code

Reply:
We talked about it, it's a small thing. Dror suggested having it as it is for debugging purposes, it's easier to see strings than bytes if/when something goes wrong